### PR TITLE
Avoid using deprecated -cp argument from Launcher

### PR DESCRIPTION
**Pull request by lgemeinhardt**
Avoid using deprecated -cp argument from Launcher and use "normal" classpath to add the additional jars. This fix enables to use Java > 8 for this method.

### Proposed changelog entries

Developer: Support Java 11 in `hudson.slaves.Channels#newJVM`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since="TODO")` or `@Deprecated(since="TODO", forRemoval=true)` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6723"></a>

### CODE REVIEW

**Review by jglick — COMMENTED** [review]
:shrug: looks harmless if it works. Adjusting the agent classpath to include something other than `agent.jar` is a bad idea, and probably `Channels.newJVM` is a bad idea as well. The stated justifications for the parameter https://github.com/jenkinsci/jenkins/blob/512f267b4cf504a2641b9a44c0c842322e72c996/core/src/main/java/hudson/slaves/Channels.java#L205-L208 sound like nonsense. I _think_ this method is used by plugins which launch an extra agent-like process on the agent computer that both runs some Jenkins plugin code and a third-party service, which is a really bad idea for multiple reasons.

**Comment by daniel-beck** [comment]
Seems it's coming from https://github.com/jenkinsci/hadoop-plugin/commit/bc9379d9965a68a9f947449f652bd42fc0492964

>  moved some of the code to the core to reuse them with the Selenium plugin

Should probably be deleted without replacement.

**Comment by jglick** [comment]
https://plugins.jenkins.io/hadoop/ is long dead.

And `createSeleniumRCVM`, which does not use the core method, looks like it will also not work on Java 11? https://github.com/jenkinsci/selenium-plugin/blob/63e2d642d480810a43c6d8bf93bfb4a6d7704567/src/main/java/hudson/plugins/selenium/process/SeleniumProcessUtils.java#L90

I am also inclined to think that `Channels.newJvm` should just be deprecated or deleted without replacement, and whoever adopts https://plugins.jenkins.io/selenium/ (and fixes the unresolved security vulnerability) can either inline and patch the impls they need as a hotfix, or just redesign the crazy usage of Remoting channels.

(`Channels.forProcess` is a key part of the Evil Plugin: https://github.com/jenkinsci/maven-plugin/blob/2f1a2ddf4a849301d594d1aa99aeb1e0cf5ec6d3/src/main/java/hudson/maven/AbstractMavenProcessFactory.java#L296 Beware also https://github.com/jenkinsci/selenium-plugin/blob/63e2d642d480810a43c6d8bf93bfb4a6d7704567/src/main/java/hudson/plugins/selenium/process/SeleniumProcessUtils.java#L104)

**Comment by jglick** [comment]
> if it works

https://github.com/jenkinsci/remoting/pull/554#issuecomment-1169983969 implies that the Selenium plugin continues to function even if the classpath is ignored completely, so I am not sure what to believe. I am not really sure why `Launcher.addClasspath` was created to begin with, rather than passing the regular Java VM classpath argument as here.

File: core/src/main/java/hudson/slaves/Channels.java
Comment (jglick) [inline core/src/main/java/hudson/slaves/Channels.java:226]: Note use of `pathSeparator` from controller seems appropriate given that https://github.com/jenkinsci/jenkins/blob/512f267b4cf504a2641b9a44c0c842322e72c996/core/src/main/java/hudson/slaves/Channels.java#L230 this is apparently run on the same machine as the controller? :scream:

**Review by jglick — APPROVED** [review]
Given https://github.com/jenkinsci/remoting/pull/468 & https://github.com/jenkinsci/jenkins/pull/6732 it seems like this is OK. Again I do not know how the Selenium plugin is actually supposed to work so I cannot evaluate whether this makes sense, but given that `-cp` does not work at all on Java 11 I do not see how changing the impl now could make anything worse.

File: core/src/main/java/hudson/slaves/Channels.java
Comment (lgemeinhardt) [reply core/src/main/java/hudson/slaves/Channels.java:226]: It is running on the same machine. It was the same "trick" before (concat and later do split) ... Now this could be avoided, if ClasspathBuilder could directly be added to another ClasspathBuilder instance. 
In general I do not know, why the author of the selenium plugin did not put all the classpath together (like now proposed), but I know this change keeps the plugin running with Java > 8.

**Comment by jglick** [comment]
#6732 was closed (without explanation) so I would recommend deprecating the two `newJVM` overloads here as a reminder that this is just a hotfix for the `selenium` plugin until it can be designed better (or itself deprecated in favor of some other idiom).

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
